Lib/typing.py

```diff
@@ -2412,10 +2414,79 @@ def utf8(value: bytes) -> bytes: ...
       def utf8(value: str) -> bytes: ...
       def utf8(value):
           # implementation goes here
+
+    The overloads for a function can be retrieved at runtime using the
+    get_overloads() function.
     """
+    # Inline version of _get_key_for_callable
+    # classmethod and staticmethod
+    func = getattr(func, "__func__", func)
+    try:
+        key = (func.__module__, func.__qualname__)
+    except AttributeError:
+        # Not a normal function; ignore.
+        pass
+    else:
+        # We're inlining get_overloads() here to avoid computing the key twice.
+        existing = _overload_registry.setdefault(key, [])
+        if existing:
+            # If we are registering a variant with a lineno below or equal to that of the
+            # most recent existing variant, we're probably re-creating overloads for a
+            # function that already exists. In that case, we clear the existing variants
+            # to avoid leaking memory.
+            firstlineno = _get_firstlineno(func)
+            if firstlineno is not None:
+                existing_lineno = _get_firstlineno(existing[-1])
+                if existing_lineno is not None and firstlineno <= existing_lineno:
+                    existing.clear()
```

Comment:
It looks like we can get a significant speedup by inlining `_get_firstlineno` as well:

```suggestion
            func2 = getattr(func, "__func__", func)
            try:
                firstlineno = func2.__code__.co_firstlineno
            except AttributeError:
                pass
            else:
                try:
                    existing_lineno = existing[-1].__code__.co_firstlineno
                except AttributeError:
                    pass
                else:
                    if firstlineno <= existing_lineno:
                        existing.clear()
```

On my machine, with `overload` being the current implementation, and `overload2` being the version with `_get_firstlineno` inlined:

```python
In [9]: def f(): ...

In [10]: %timeit overload(f)
700 ns ± 1.53 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [11]: %timeit overload2(f)
465 ns ± 7.56 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

Reply:
What kind of machine is that, @AlexWaygood? On my machine (a 3 year old x86 MacBook Pro) the original `overload(f)` takes 50 nsec. Your machine is 30x faster?!

Reply:
Sorry -- by "current implementation", I meant "current implementation suggested in this PR". And those timings I posted above are in nanoseconds, so that means that Jelle's proposed implementation takes 700 nsec, whereas with my revisions, it would take 522 nsec. So both would still be markedly slower than 50 nsec, rather than 30x faster :)

Reply:
Anyway, my machine is a lenovo Windows laptop, which is fairly new but nothing special.

Reply:
I am trying out a slight variation this locally now, along with the len() < 16 idea from the other thread. Now the overhead is down to ~7x:

```
In [3]: %timeit overload(f)
294 ns ± 1.85 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [6]: %timeit old_overload(f)
44 ns ± 0.256 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [7]: 294 / 44
Out[7]: 6.681818181818182
```

Reply:
Where are we at after changing to use a `defaultdict` instead of a `dict`?

Reply:
I got `472 ns ± 11.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)`, but I backed out the len()>16 check for now pending Guido's response on the other thread.

Reply:
> Sorry for my lack of clarity -- by "current implementation", I meant "current implementation suggested in this PR". And those timings I posted above are in nanoseconds, so that means that Jelle's proposed implementation takes 700 nsec, whereas with my revisions, it would take 522 nsec. So both would still be markedly slower than 50 nsec, rather than 30x faster :)

D'oh, re-reading your post I realize I mistook the stdev for the mean value. :-)

I suppose we shouldn't overly worry about this slowdown, but I think it might be worth asking on typing-sig if anyone would actually mind `@overload` getting 10x slower.